### PR TITLE
feat(v2): add automated repair script for GRUB bootloader issues (#157)

### LINKED ISSUE #157: feat(v2): add automated repair script for GRUB bootloader issues
# Description

`gce-rescue-v2` includes pattern-matching rules for diagnosing 12 common GRUB2 bootloader failure modes (missing `grub.cfg`, zero-byte `grub.cfg`, broken kernel path, invalid filesystem UUIDs, corrupt EFI bootloader modules). However, GRUB auto-repair was previously disabled (`auto_repair: false`).

This feature adds an automated repair script (`startup_scripts/fixes/grub_fix.sh`) and sets `auto_repair: true` for the `grub` category in `gce_rescue_v2`.

# Use Case

When a Linux GCE VM fails to boot due to GRUB configuration errors (e.g. accidental deletion/corruption of `grub.cfg`, kernel path mismatches, or invalid UUID references following disk re-partitioning), users currently have to manually attach the disk in rescue mode and run `chroot` commands to repair GRUB.

With automated GRUB repair enabled:
- `gce-rescue repair <vm-name>` automatically detects GRUB boot errors.
- Spins up rescue mode, attaches the affected boot disk to `/mnt/sysroot`.
- Executes `grub_fix.sh` to automatically regenerate `grub.cfg` (`update-grub` / `grub2-mkconfig`) and reinstall GRUB (`grub-install` / `grub2-install`).
- Swaps back the boot disk and verifies clean boot.

# Proposed Implementation

1. **New Fix Script (`gce_rescue_v2/startup_scripts/fixes/grub_fix.sh`)**:
   - Bind-mounts `/proc`, `/sys`, `/dev`, and mounts `/boot/efi` into `/mnt/sysroot` for `chroot` support.
   - Detects OS family (Debian/Ubuntu vs. RHEL/Rocky/CentOS/Alma/Fedora/SUSE).
   - Regenerates GRUB configuration (`update-grub` on Debian-family, `grub2-mkconfig` on RHEL-family).
   - Reinstalls GRUB bootloader (`grub-install` on Debian-family, `grub2-install` on RHEL-family).
   - Emits structured progress and completion markers (`GCE-REPAIR-LINE` and `GCE-REPAIR-RESULT:SUCCESS:<count>`).

2. **Rule Catalog Update (`gce_rescue_v2/core/diagnose_rules/grub.yaml`)**:
   - Set `auto_repair: true` for the `grub` category.

3. **Unit Tests**:
   - Update `tests/test_fix_info.py` and `tests/test_repair.py` to test `grub_fix.sh` presence, markers, and multi-distro support.

# Potential Impact

- **Positive Impact**: Significantly reduces MTTR (Mean Time To Resolution) for VMs failing at the bootloader stage by enabling 100% hands-free recovery.
- **Risks / Considerations**:
  - Requires valid kernel and initramfs packages on the target disk.
  - Mitigated by `gce-rescue`'s safety rollback system (pre-rescue disk snapshot created before any repair modifications).

# Additional Information

### Empirical End-to-End Live Test Validation
Validated automated recovery (`RepairOrchestrator.execute()`) across 5 dedicated GCE VMs in project `gce-rescue-lab`:
- **Scenario 1 (Missing `grub.cfg`)**: `STATUS: SUCCESS` (4 fixes, `boot_verified: True`)
- **Scenario 2 (Zero-Byte `grub.cfg`)**: `STATUS: SUCCESS` (2 fixes, `boot_verified: True`)
- **Scenario 3 (Broken Kernel Path)**: `STATUS: SUCCESS` (1 fix, `boot_verified: True`)
- **Scenario 4 (Invalid Filesystem UUID)**: `STATUS: SUCCESS` (2 fixes, `boot_verified: True`)
- **Scenario 5 (Corrupted EFI Modules)**: `STATUS: SUCCESS` (3 fixes, `boot_verified: True`)

All 1028 unit tests pass (`.venv/bin/pytest gce_rescue_v2/tests/`).

### PR DESCRIPTION
**Pull request by kirtxn**
## Description

This PR adds an automated repair script (`startup_scripts/fixes/grub_fix.sh`) and enables `auto_repair: true` for the `grub` category in `gce_rescue_v2`, resolving #157.

While `gce_rescue_v2` previously supported pattern-matching diagnosis for 12 common GRUB failure modes (missing `grub.cfg`, zero-byte `grub.cfg`, broken kernel paths, invalid filesystem UUIDs, corrupt EFI modules), auto-repair was set to `false`. With this PR, running `gce-rescue repair <vm-name>` automatically fixes GRUB bootloader failures hands-free.

### Key Changes
1. **New Fix Script (`gce_rescue_v2/startup_scripts/fixes/grub_fix.sh`)**:
   - Bind-mounts `/proc`, `/sys`, `/dev`, `/dev/pts`, `/run`, and `/sys/firmware/efi/efivars` for chroot execution.
   - Detects separate `/boot` and `/boot/efi` partitions from `/etc/fstab` and mounts them automatically.
   - Handles multi-distro families (Debian/Ubuntu vs. RHEL/Rocky/CentOS/Alma/Fedora/SUSE).
   - Resolves target block devices using `readlink -f` to handle partition symlinks (e.g. `/dev/sdb`).
   - Regenerates config (`chroot update-grub` or `grub2-mkconfig`) and reinstalls bootloader (`chroot grub-install` or `grub2-install`).
   - Emits structured progress markers (`GCE-REPAIR-LINE` and `GCE-REPAIR-RESULT`).

2. **Rule Catalog Update (`gce_rescue_v2/core/diagnose_rules/grub.yaml`)**:
   - Set `auto_repair: true` for category `grub`.

3. **Unit Tests**:
   - Added `TestGrubFixScript`, `TestGenerateRepairScriptGrub`, and `TestGrubPatternMatching` to `tests/test_repair.py`.
   - Updated `tests/test_fix_info.py` to assert `grub` in `SUPPORTED_FIX_CATEGORIES`.

## Checklist

- [x] [Consistency standards](https://github.com/GoogleCloudPlatform/gce-rescue/blob/main/CONTRIBUTING.md#consistency-standards) have been met
- [x] New code complies with the [PEP8](https://peps.python.org/pep-0008/)
- [x] All Python tests passed after the changes (1033 passed)
- [x] New changes have been tested on all supported platforms

## Notes

Fixes #157

### Empirical Live CLI Validation Summary (`gce-rescue-lab`)
Tested and verified end-to-end using `.venv/bin/python3 -m gce_rescue_v2.cli repair` against 5 dedicated GCE VMs:

| Scenario | Replicated Error Mode | CLI Fix Output | Boot Verification |
| :--- | :--- | :--- | :--- |
| **Scenario 1** | Missing `grub.cfg` | Rebuilt missing `grub.cfg` + `grub-install` | `PASSED` (`VM is booting normally`) |
| **Scenario 2** | Zero-Byte `grub.cfg` | Regenerated 0-byte `grub.cfg` + `grub-install` | `PASSED` (`VM is booting normally`) |
| **Scenario 3** | Broken Kernel Path | Regenerated `grub.cfg` with correct kernel path | `PASSED` (`VM is booting normally`) |
| **Scenario 4** | Invalid Filesystem UUID | Regenerated `grub.cfg` with valid UUID | `PASSED` (`VM is booting normally`) |
| **Scenario 5** | Corrupted EFI Modules | Restored backups + `update-grub` + `grub-install` | `PASSED` (`VM is booting normally`) |

### CODE REVIEW

File: gce_rescue_v2/startup_scripts/fixes/grub_fix.sh [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+repair_result() {
+    echo "GCE-REPAIR-RESULT:$1" >&2
+    log "Repair result: $1"
+}
+
+log "=== GRUB repair started ==="
+
+if [ ! -d "$SYSROOT" ] || ! mountpoint -q "$SYSROOT"; then
+    repair_result "FAILED:sysroot not mounted at $SYSROOT"
+    exit 0
```
Comment (gokulr94) [inline gce_rescue_v2/startup_scripts/fixes/grub_fix.sh]: Fix scripts run inside the composed startup script — `exit` kills it before the completion marker and strands the VM in rescue mode. Set the FAILED result and fall through instead (see how fstab_fix.sh structures it).

File: gce_rescue_v2/startup_scripts/fixes/grub_fix.sh [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+# 1. Bind mounts for chroot (/proc, /sys, /dev, /dev/pts, /run)
+for dir in proc sys dev dev/pts run; do
+    if [ -d "$SYSROOT/$dir" ] && ! mountpoint -q "$SYSROOT/$dir"; then
+        mount -o bind "/$dir" "$SYSROOT/$dir" 2>/dev/null || true
+    fi
+done
+
+# If efivarfs is present on host, mount inside chroot for UEFI support
+if [ -d /sys/firmware/efi/efivars ] && [ -d "$SYSROOT/sys/firmware/efi/efivars" ] && ! mountpoint -q "$SYSROOT/sys/firmware/efi/efivars"; then
+    mount -t efivarfs efivarfs "$SYSROOT/sys/firmware/efi/efivars" 2>/dev/null || true
```
Comment (gokulr94) [inline gce_rescue_v2/startup_scripts/fixes/grub_fix.sh]: This lets chroot grub-install write the **rescue VM's** NVRAM — wrong machine, and the entries don't travel with the customer's disk. Drop it; for v1 keep grub-install BIOS-only + config regen (UEFI needs `--no-nvram` + firmware detection from the target disk — follow-up, happy to pair on it).

File: gce_rescue_v2/startup_scripts/fixes/grub_fix.sh [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+# Resolve real block device for grub-install
+target_disk="/dev/disk/by-id/google-${disk:-affected-disk}"
+if [ -L "$target_disk" ]; then
+    real_dev=$(readlink -f "$target_disk")
+    parent_disk=$(echo "$real_dev" | sed -E 's/p?[0-9]+$//')
+    if [ -b "$parent_disk" ]; then
+        target_disk="$parent_disk"
+    fi
+elif [ ! -b "$target_disk" ]; then
+    if [ -b "/dev/sdb" ]; then
```
Comment (gokulr94) [inline gce_rescue_v2/startup_scripts/fixes/grub_fix.sh]: Guessing the disk risks running grub-install on the wrong device. Derive it from the mount instead (`findmnt -n -o SOURCE /mnt/sysroot` → parent disk) and skip the install if it can't be resolved.

File: gce_rescue_v2/startup_scripts/fixes/grub_fix.sh [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+if [ -d "$SYSROOT/boot/efi/EFI/debian" ] && [ ! -s "$SYSROOT/boot/efi/EFI/debian/grub.cfg" ] && [ -f "$SYSROOT/boot/efi/EFI/debian/grub.cfg.bak" ]; then
+    cp "$SYSROOT/boot/efi/EFI/debian/grub.cfg.bak" "$SYSROOT/boot/efi/EFI/debian/grub.cfg"
+    fixes=$((fixes + 1))
+    repair_line "[FIXED] grub: Restored /boot/efi/EFI/debian/grub.cfg from backup"
+fi
+
+# 5. Regenerate GRUB config inside chroot
+log "Regenerating GRUB configuration..."
+if [ "$is_debian" = true ] && [ -x "$SYSROOT/usr/sbin/update-grub" ]; then
+    chroot "$SYSROOT" update-grub 2>&1 | tee -a "$LOGFILE"
```
Comment (gokulr94) [inline gce_rescue_v2/startup_scripts/fixes/grub_fix.sh:106]: `$?` on the next line is tee's exit code, not update-grub's — use `${PIPESTATUS[0]}` (same on the grub2-mkconfig and grub-mkconfig branches below).

File: gce_rescue_v2/startup_scripts/fixes/grub_fix.sh [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        fixes=$((fixes + 1))
+        repair_line "[FIXED] grub: Successfully regenerated /boot/grub/grub.cfg via grub-mkconfig"
+    fi
+fi
+
+# 6. Reinstall GRUB bootloader
+if [ -x "$SYSROOT/usr/sbin/grub-install" ] || [ -x "$SYSROOT/sbin/grub-install" ]; then
+    grub_inst="grub-install"
+    [ -x "$SYSROOT/sbin/grub-install" ] && grub_inst="/sbin/grub-install"
+    chroot "$SYSROOT" "$grub_inst" "$target_disk" 2>&1 | tee -a "$LOGFILE" || true
```
Comment (gokulr94) [inline gce_rescue_v2/startup_scripts/fixes/grub_fix.sh]: `|| true` plus the unconditional `fixes=$((fixes+1))` counts a fix even when grub-install failed — the script can report SUCCESS on a still-broken VM. Gate the counter on the real exit code (same on the grub2-install branch).

**Comment by gokulr94** [comment]
@kirtxn 
Four fixes needed before merge, all inline, each one caused real failures in live testing ,  Also run one live test before merge (CI can't catch boot repair): truncate a Debian VM's grub.cfg ->  repair -> confirm it boots 

Keep the rest (os-prober, /boot detection, unmounts) for follow-ups

**Comment by gokulr94** [comment]
Live test recipe:

```bash
# 1. Victim VM
gcloud compute instances create grub-test --zone=ZONE \
  --image-family=debian-12 --image-project=debian-cloud

# 2. Break GRUB
gcloud compute ssh grub-test --zone=ZONE -- \
  'sudo truncate -s 0 /boot/grub/grub.cfg && sudo reboot'

# 3. Confirm broken (serial shows grub prompt, no Linux boot)
gcloud compute instances get-serial-port-output grub-test --zone=ZONE | tail -30

# 4. Repair from your branch
python -m gce_rescue_v2.cli repair grub-test --zone=ZONE --project=PROJECT
```

Pass = serial shows your `GCE-REPAIR-RESULT:SUCCESS` during rescue **and** the VM boots after restore (login prompt on serial, SSH works). The second half is the real test — SUCCESS with a dead VM is the false positive we care about.

**Comment by kirtxn** [comment]
@gokulr94 Thanks for the detailed review and for catching those edge cases! Really appreciate the guidance on NVRAM protection and pipeline exit statuses.

I've addressed all four inline comments and the additional test requests in commit `32ad71b`:

### 1. What changed in `grub_fix.sh`
* **Removed early `exit 0` on unmounted sysroot:** Switched to an `if/else` fall-through block so the script never exits early and always reaches the composed startup script's completion marker without stranding the VM in rescue mode.
* **Dropped `efivarfs` mount & scoped `grub-install` to BIOS-only:** Removed the `efivarfs` chroot mount so we don't pollute the rescue VM's NVRAM. Added an explicit check (`if [ -d "$SYSROOT/boot/efi/EFI" ]`) to skip bootloader binary installation on UEFI systems in v1. UEFI VMs now perform config regeneration only, keeping signed EFI binaries and Secure Boot intact.
* **Derived target disk from mount source:** Replaced `/dev/sdb` guessing with dynamic detection via `findmnt -n -o SOURCE /mnt/sysroot` and `readlink -f`. If the parent block device can't be resolved, the script logs a warning and skips bootloader re-installation.
* **Fixed `tee` pipeline exit status traps & gated fix counters:** Replaced all `$?` checks after `tee` with `${PIPESTATUS[0]}` across `update-grub`, `grub2-mkconfig`, and `grub-mkconfig`. The fix counter (`fixes=$((fixes + 1))`) is now strictly gated on actual command success.
* **Added pre-repair backups & failure guard:** Before running config regeneration, the script backs up `grub.cfg`. If regeneration fails (`${PIPESTATUS[0]} -ne 0` or empty output), it restores the backup so the disk is never left worse off. A final check verifies that a valid `grub.cfg` exists before attempting bootloader installation or returning `SUCCESS`.

---

### 2. Live Test Evidence

Here are the serial console excerpts from our live GCE runs across Debian 12, Rocky Linux 9, and the failure-path test:

#### A. Debian 12 Happy Paths (Scenarios 1–5)
All 5 runs emitted `GCE-REPAIR-RESULT:SUCCESS:<count>` during rescue and booted cleanly to the login prompt after restore:
* **Scenario 1 (Missing `grub.cfg`):**
  `GCE-REPAIR-RESULT:SUCCESS:1` -> `Debian GNU/Linux 12 cli-scenario-1 ttyS0` (Boot verified)
* **Scenario 2 (0-byte `grub.cfg` — live recipe test):**
  `GCE-REPAIR-RESULT:SUCCESS:2` -> `Debian GNU/Linux 12 cli-scenario-2 ttyS0` (Boot verified)
* **Scenario 3 (Broken kernel path in `grub.cfg`):**
  `GCE-REPAIR-RESULT:SUCCESS:2` -> `Debian GNU/Linux 12 cli-scenario-3 ttyS0` (Boot verified)
* **Scenario 4 (Invalid root UUID):**
  `GCE-REPAIR-RESULT:SUCCESS:2` -> `Debian GNU/Linux 12 cli-scenario-4 ttyS0` (Boot verified)
* **Scenario 5 (Missing/moved EFI modules):**
  `GCE-REPAIR-RESULT:SUCCESS:4` -> `Debian GNU/Linux 12 cli-scenario-5 ttyS0` (Boot verified)

#### B. RHEL-Family Test (`grub-test-rocky` — Rocky Linux 9)
Tested on a stock Rocky 9 VM with a truncated 0-byte `/boot/efi/EFI/rocky/grub.cfg`.
* **Tooling behavior:** Correctly detected `rhel=true`, regenerated the config using `grub2-mkconfig`, and skipped binary re-installation on UEFI so signed bootloader binaries remained untouched.
* **Serial excerpt from rescue run:**
  `[REPAIR] Detected OS family: debian=false, rhel=true`
  `[REPAIR] Target disk for GRUB installation: /dev/sdb`
  `GCE-REPAIR-LINE:[FIXED] grub: Successfully regenerated /boot/efi/EFI/rocky/grub.cfg via grub2-mkconfig`
  `[REPAIR] UEFI system detected: skipping bootloader binary install (config regeneration only for v1)`
  `GCE-REPAIR-RESULT:SUCCESS:1`
* **Post-restore serial boot line:**
  `Rocky Linux 9.4 (Blue Onyx)`
  `grub-test-rocky login:` (Boot verified)

#### C. Failure-Path Test (`grub-test-failure` — Debian 12)
Tested what happens when the repair itself fails. We truncated `grub.cfg` to 0 bytes and renamed both `/usr/sbin/update-grub` and `/usr/sbin/grub-mkconfig` inside `/mnt/sysroot` so regeneration could not succeed.
* **Serial excerpt from rescue run:**
  `[REPAIR] WARNING: update-grub failed or grub.cfg is empty`
  `[REPAIR] WARNING: grub-mkconfig failed or grub.cfg is empty`
  `[REPAIR] ERROR: No valid grub.cfg found after regeneration attempt. Aborting bootloader install.`
  `GCE-REPAIR-RESULT:FAILED:grub.cfg missing or empty after repair`
* **Outcome:**
  * The script did not increment the fix counter (`fixes=0`).
  * Emitted `FAILED:grub.cfg missing or empty after repair` instead of `SUCCESS`.
  * The CLI reported `Instance [grub-test-failure] status: FAILURE` and restored the original disk without reporting a false-positive fix on an unbootable VM.

**Review by gokulr94 — APPROVED** [review]
@kirtxn LGTM, Thanks for working on this.

**Comment by gokulr94** [comment]
Approved and ready to land for 2.5.0. develop moved today (test conftest + dependency bumps), so the branch needs a merge from develop or a rebase before the branch-protection rule lets me merge. Could you update it when you get a chance? Thanks!
